packages/sdk/src/oracle/client.ts

```diff
@@ -1,87 +1,33 @@
+import assert from "assert";
 import Store, { Emit } from "./store";
 import type { state } from "./types";
 import { Inputs } from "./types/state";
 import { Signer } from "./types/ethers";
-import * as utils from "./utils";
+import { Update } from "./services/update";
+import { StateMachine } from "./services/statemachines/statemachine";
+import { loop } from "../utils";
 
 export class Client {
-  public readonly update: Update;
-  constructor(public readonly store: Store) {
-    this.update = new Update(store);
-  }
-  setUser(address: string, chainId: number, signer: Signer): void {
-    this.store.write((write) => write.user().set({ address, chainId, signer }));
-  }
-  setActiveRequest(params: Inputs["request"]): void {
-    const { requester, identifier, timestamp, ancillaryData, chainId } = params;
-    this.store.write((write) => write.inputs().request(requester, identifier, timestamp, ancillaryData, chainId));
-  }
-  previewProposal() {
-    return utils.previewProposal(this.store.get());
-  }
-}
-
-export class Update {
-  private read: Store["read"];
-  private write: Store["write"];
-  constructor(private store: Store) {
-    this.read = store.read;
-    this.write = store.write;
-  }
-  async all() {
-    await this.oracle();
-    await this.request();
-    await this.collateralProps();
-    await this.userCollateralBalance();
-    await this.oracleAllowance();
-  }
-  async request() {
-    const request = this.read().inputRequest();
-    const chainId = request.chainId;
-    const oo = this.read().oracleService();
-    const fullRequest = await oo.getRequest(
-      request.requester,
-      request.identifier,
-      request.timestamp,
-      request.ancillaryData
-    );
-    const state = await oo.getState(request.requester, request.identifier, request.timestamp, request.ancillaryData);
-    this.write((write) => {
-      // create the erc20 service to handle currency
-      write.services(chainId).erc20s(fullRequest.currency);
-      write
-        .chains(chainId)
-        .optimisticOracle()
-        .request(request, { ...fullRequest, state });
+  private intervalStarted = false;
+  constructor(public readonly store: Store, public readonly update: Update, public readonly sm: StateMachine) {}
+  setUser(address: string, chainId: number, signer: Signer): string {
+    return this.sm.types.setUser.create({ address, chainId, signer });
+  }
+  setActiveRequest(params: Inputs["request"]): string {
+    return this.sm.types.setActiveRequest.create(params);
+  }
+  // runs statemachine step loop pretty fast by default.
+  startInterval(delayMs = 10) {
```

Comment:
this ensures the statemachine queue is constantly being checked. The speed doesnt really matter that much, but we do want something pretty reactive to input.  in a perfect world this would run from events rather than a timing loop.

Reply:
it would just be a more complicated for some gain in performance. The contexts have this "sleep" feature which means even if you had an event driven system you still need a continuous loop to know when to "wake" the contexts to re-run, so probably not even worth persuing at this point